### ISSUE TITLE
What do parent and key do in this code?

### ISSUE THREAD

**Issue by dstathis** [canonical/operator#610]
https://github.com/canonical/operator/blob/2875e73e3cacb3b0aa9bbf96524b9b4e7f23b67f/ops/framework.py#L332

I don't really understand what these arguments are for. What do they do? Maybe we could add a docstring?

**Comment by petevg** [comment]
I believe that the purpose of the code is to keep a reference to the “framework” object, which may or may not be the thing that we’re subclassing. That enables charm authors to call self.framework on classes like Charm, Model and Application, without worrying about tracking how the framework is instantiated, or worrying about duplication.

Documentation would do absolutely no harm, of course :-)

~ PeteVG

launchpad.net/~petevg

**Comment by dstathis** [comment]
Yeah I've more or less figured this out for parent. Key is the one that is more of a mystery to me

**Comment by jameinel** [comment]
I think expanding the documentation is a great idea. I'll give a little bit
of where that might go.

Objects in OF must be uniquely identified, and if you ever have 2 with the
same handle, you *should* get an exception letting you know.

The reason for this is because event handlers, etc will send events to a
named object, but if you have multiple instances, what is actually
registered to that name. Going further, it also helps to guarantee the
stability of charms.
When a charm evaluates an event, juju has create a python process, handed
it a bit of environment variables for it to know what event is supposed to
be happening, and then the Operator Framework instantiates the state of the
world (what other events are deferred, what stored state has been made
available, etc).
But that process only exists for the lifetime of that hook. So to be able
to restore the state of the world on the *next* hook invocation, we want
clear identifiers for the Objects in play.
This is what makes it possible to write code that feels like one event
happens and the next event happens, even though all the context is built up
and torn down between each hook.

Is that helpful in understanding why we have a key and why we consider it
required?

**Comment by jnsgruk** [comment]
Nice context, @jameinel.

@petevg seems like we should get this into a relevant docstring somewhere?

**Comment by petevg** [comment]
Thanks, @jameinel! @jnsgruk I'll make a task and push a commit encapsulating the above.
